### ISSUE TITLE
Improve LSTM ML model

### ISSUE THREAD

**Issue by tyler-pruitt** [alpha-b0t/AlphaB0t#104]
Me: Build me an LSTM ML model for pricing the Hang Seng index in python

Chatgpt:
Below is a clean, end-to-end example of an LSTM model in Python for forecasting the Hang Seng Index (HSI) using historical price data.
It’s written so you can run it locally with minimal modification.

I’ll assume:
	•	You want to predict the next closing price
	•	You are comfortable with Keras / TensorFlow
	•	Data is daily historical prices

⸻

1. Install dependencies

pip install numpy pandas scikit-learn matplotlib tensorflow yfinance

⸻

2. Load Hang Seng Index data

Yahoo Finance ticker for Hang Seng Index:

^HSI

import numpy as np
import pandas as pd
import yfinance as yf
from sklearn.preprocessing import MinMaxScaler

# Download historical data
df = yf.download("^HSI", start="2000-01-01", end="2025-01-01")

# Use closing prices only
data = df[['Close']].dropna()

⸻

3. Normalize data

LSTMs work best on scaled data.

scaler = MinMaxScaler(feature_range=(0, 1))
scaled_data = scaler.fit_transform(data)

⸻

4. Create time-series sequences

We’ll use 60 days to predict the next day.

def create_sequences(data, window_size=60):
    X, y = [], []
    for i in range(window_size, len(data)):
        X.append(data[i-window_size:i, 0])
        y.append(data[i, 0])
    return np.array(X), np.array(y)

X, y = create_sequences(scaled_data)

Reshape for LSTM:

X = np.reshape(X, (X.shape[0], X.shape[1], 1))

⸻

5. Train / test split

split = int(0.8 * len(X))
X_train, X_test = X[:split], X[split:]
y_train, y_test = y[:split], y[split:]

⸻

6. Build the LSTM model

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout

model = Sequential()

model.add(LSTM(50, return_sequences=True, input_shape=(X_train.shape[1], 1)))
model.add(Dropout(0.2))

model.add(LSTM(50, return_sequences=False))
model.add(Dropout(0.2))

model.add(Dense(1))

model.compile(
    optimizer='adam',
    loss='mean_squared_error'
)

model.summary()

⸻

7. Train the model

history = model.fit(
    X_train, y_train,
    epochs=25,
    batch_size=32,
    validation_split=0.1
)

⸻

8. Make predictions

predictions = model.predict(X_test)

# Inverse scale
predictions = scaler.inverse_transform(predictions.reshape(-1, 1))
y_test_actual = scaler.inverse_transform(y_test.reshape(-1, 1))

⸻

9. Plot results

import matplotlib.pyplot as plt

plt.figure(figsize=(12,6))
plt.plot(y_test_actual, label="Actual HSI")
plt.plot(predictions, label="Predicted HSI")
plt.title("Hang Seng Index Price Prediction (LSTM)")
plt.xlabel("Time")
plt.ylabel("Index Value")
plt.legend()
plt.show()

⸻

10. Predict next day price

last_60_days = scaled_data[-60:]
last_60_days = last_60_days.reshape(1, 60, 1)

next_price_scaled = model.predict(last_60_days)
next_price = scaler.inverse_transform(next_price_scaled)

print("Next predicted HSI price:", next_price[0][0])

⸻

Model limitations (important)
	•	❌ Predicts price direction poorly during regime changes
	•	❌ No macro factors (rates, FX, China policy)
	•	❌ LSTM ≠ trading strategy
	•	✔️ Good for trend smoothing and volatility learning

⸻

Want upgrades?

I can:
	•	Add technical indicators (RSI, MACD, volatility)
	•	Convert to returns instead of price
	•	Add attention / Transformer
	•	Build a walk-forward backtest
	•	Turn this into a trading strategy

Just tell me what level you want 🚀

Me: Why is your test data only at the end? Why does your training data only have close instead of open, high, etc.? Then do the five bullets

Chatgpt: Rest of chat (https://chatgpt.com/share/696fed6c-1c90-8010-8e30-ecbd4e6d79e2)
